staging/src/k8s.io/client-go/tools/auth/exec/exec.go

```diff
@@ -0,0 +1,147 @@
+}
+
+// LoadExecCredentialFromEnv is a helper-wrapper around LoadExecCredential that loads from the
+// well-known KUBERNETES_EXEC_INFO environment variable.
+//
+// When the KUBERNETES_EXEC_INFO environment variable is not set or is empty, then this function
+// will immediately return an error.
+func LoadExecCredentialFromEnv() (runtime.Object, *rest.Config, error) {
+	env := os.Getenv(execInfoEnv)
+	if env == "" {
+		return nil, nil, errors.New("KUBERNETES_EXEC_INFO env var is unset or empty")
+	}
+	return LoadExecCredential([]byte(env))
+}
+
+// LoadExecCredential loads the configuration needed for an exec plugin to communicate with a
+// cluster.
+//
+// LoadExecCredential expects the provided data to be a serialized client.authentication.k8s.io
+// ExecCredential object (of any version). If the provided data is invalid (i.e., it cannot be
+// unmarshalled into any known client.authentication.k8s.io ExecCredential version), an error will
+// be returned. A successfully unmarshalled ExecCredential will be returned as the first return
+// value.
+//
+// If the provided data is successfully unmarshalled, but it does not contain cluster information
+// (i.e., ExecCredential.Spec.Cluster == nil), then the returned rest.Config and error will be nil.
+//
+// Note that the returned rest.Config will use anonymous authentication, since the exec plugin has
+// not returned credentials for this cluster yet.
+func LoadExecCredential(data []byte) (runtime.Object, *rest.Config, error) {
+	obj, gvk, err := codecs.UniversalDeserializer().Decode(data, nil, nil)
+	if err != nil {
+		return nil, nil, fmt.Errorf("decode: %w", err)
+	}
+
+	expectedGK := schema.GroupKind{
+		Group: clientauthentication.SchemeGroupVersion.Group,
+		Kind:  "ExecCredential",
+	}
+	if gvk.GroupKind() != expectedGK {
+		return nil, nil, fmt.Errorf(
+			"invalid group/kind: wanted %s, got %s",
+			expectedGK.String(),
+			gvk.GroupKind().String(),
+		)
+	}
+
+	// Explicitly convert object here so that we can return a nicer error message above for when the
+	// data represents an invalid type.
+	var execCredential clientauthentication.ExecCredential
+	if err := scheme.Convert(obj, &execCredential, nil); err != nil {
+		return nil, nil, fmt.Errorf("cannot convert to ExecCredential: %w", err)
+	}
+
+	if execCredential.Spec.Cluster == nil {
+		return nil, nil, errors.New("exec credential does not contain cluster information")
+	}
+
+	// make a kubeconfig first, then translate that to a rest.Config.
+	kubeconfig := makeKubeconfig(execCredential.Spec.Cluster)
+	restConfig, err := clientcmd.NewNonInteractiveClientConfig(
+		*kubeconfig,
+		kubeconfig.CurrentContext,
+		&clientcmd.ConfigOverrides{},
+		nil,
+	).ClientConfig()
+	if err != nil {
+		return nil, nil, fmt.Errorf("load rest config: %w", err)
+	}
+
+	return obj, restConfig, nil
+}
+
+// makeKubeconfig makes a kubeconfig out of the provided cluster.
+func makeKubeconfig(cluster *clientauthentication.Cluster) *clientcmdapi.Config {
```

Comment:
this bit is a little awkward, but I'm not sure we have a better way to unify the two structs... is there a fuzzing test that ensures any new fields added to Cluster will have to be added here and round-tripped correctly?

Reply:
This PR adds `"k8s.io/client-go/pkg/apis/clientauthentication/v1beta1".TestCluster` which ensures that `clientauthenticationv1beta1.Cluster` has all the same fields and tags as `clientcmdv1.Cluster`, but we don't have any fuzzing test to ensure we are copying over the fields we want here in `makeKubeconfig`. So I can add a fuzzing test for that in this (`k8s.io/client-go/tools/auth/exec`) package.

Reply:
https://github.com/kubernetes/kubernetes/pull/95489/commits/bd4af0df52b7da2a838795ccf9a8760a0316da48

See my comment above about unit testing strategy...